### PR TITLE
[5.1] Only store intended redirect URL if the originating request is GET or HEAD

### LINKED ISSUE #11072: `url.intended` saves POST request url and thows 404 after login
If a guest tries to POST to a authenticated page, `url.inteded` is saved to that POST url. After login when attempting to redirect, it makes a GET request to that POST-only url and throw a NotFoundHttpException. IMHO, if current request is GET then use [`$this->generator->full()`](https://github.com/laravel/framework/blob/076a054b6c78888acb3d0b93030402dbf736d824/src/Illuminate/Routing/Redirector.php#L83) and if non-GET then it should be `$session->previousUrl()`

### PR DESCRIPTION
**Pull request by thepsion5**
This will resolve #11072, wherein a POST request to an authenticated page is the target URL of the post-authentication redirect. Now, it should redirect to the URL that first created the POST request (a resource's edit page, for example).

### CODE REVIEW

**Comment by taylorotwell** [comment]
I'm curious how a POST request would be made without first making a GET request that would have redirected?

**Comment by taylorotwell** [comment]
Do you have this problem in your own application or are you just working off of the bug report?

**Comment by thepsion5** [comment]
Not in some time, but typically it would happen when submitting a form to a POST-only route after the user's session expired. For example:

1. A user logs in and navigates to a resource index page

2. They leave the computer, their session expires

3. They return, submit a modal form on the index page to an `update-status` POST route (which does not have a GET equivalent)

4. They are redirected to the login page

5. They log back in, but are redirected to the POST-only URL

For the application in question, the update status logic had different business rules compared to a typical update and was better served as its own set of methods.

**Comment by thepsion5** [comment]
But I have run into the issue myself, yes.

**Comment by tomschlick** [comment]
> They log back in, but are redirected to the POST-only URL

Have had this happen before as well, very hard for end-users to explain and unless you're looking for it, hard to figure out why in an exception handler.

Ideally we would capture the POST input into the session, allow the user to login, and then redirect to the form with the data already filled in just like validation does.

**Comment by hassansin** [comment]
I had this problem in my application. I have a shopping cart where both guests and users can add items. But before checkout, the guest needs to login. So what I have is a `cart` resource and I show the cart using `GET /cart` for both type of users. But the checkout form has a POST-only url  like `POST /cart/checkout`.  After login it makes a GET request to that url. Instead I needed them to redirect to `GET /cart` only. I can't use the `POST /cart` for checkout as I use that endpoint to add items to the cart.

File: src/Illuminate/Routing/Redirector.php [hunk 1]
```diff
@@ -70,7 +70,8 @@ public function refresh($status = 302, $headers = [])
     }
 
     /**
-     * Create a new redirect response, while putting the current URL in the session.
+     * Create a new redirect response and stores the current URL in the session if the
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Redirector.php]: Please revert this change, and add another sentence as a "long description".

Comment (GrahamCampbell) [inline src/Illuminate/Routing/Redirector.php]:
```php
/**
 * Create a new redirect response, while putting the current URL in the session.
 *
 * Note that we're only going to store the current URL if the request was "safe".
 *
```

**Comment by GrahamCampbell** [comment]
I think this change should go to 5.2 too.

**Comment by GrahamCampbell** [comment]
It's pretty breaking.

**Comment by tomschlick** [comment]
I agree on the 5.2 point. It's changing the default behavior from what people expect.

**Comment by GrahamCampbell** [comment]
Please send to 5.2. ;)
